beluga_ros/test/test_point_cloud.cpp

```diff
@@ -0,0 +1,69 @@
+// Copyright 2023 Ekumen, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gtest/gtest.h>
+
+#include <memory>
+#include <vector>
+
+#include <range/v3/range/conversion.hpp>
+
+#include "beluga_ros/point_cloud.hpp"
+#include "beluga_ros/messages.hpp"
+
+namespace {
+
+auto make_message() {
+#if BELUGA_ROS_VERSION == 2
+  return std::make_shared<beluga_ros::msg::PointCloud2>();
+#elif BELUGA_ROS_VERSION == 1
+  return boost::make_shared<beluga_ros::msg::PointCloud2>();
+#endif
+}
+
+TEST(TestPointCloud, PointsFromUnorderedPointCloudMessage) {
```

Comment:
@pvela2017 we should also test how the interface behaves when:

- the pointcloud is empty
- the pointcloud is not 3D
- the pointcloud is not `float`

Reply:
Empty cloud seems not to be a problem if the memory size is allocated. It returns 0 values 
![image](https://github.com/user-attachments/assets/3ccd4a13-f28f-4b51-813e-8dc46b700cad)

```cpp
TEST(TestPointCloud, XYZIPointsEmptyFromUnorderedPointCloudMessage) {
  auto message = make_message();
  const auto origin = Sophus::SE3d{};
  message->width = 1;  // Unordered point cloud
  message->height = 4;  // Number of points
  // Set the point fields to x, y, z and intensity
  int fields = 4;
  message->fields.clear();
  message->fields.reserve(fields);
  int offset = 0;
  offset = addPointField(*message, "x", 1, beluga_ros::msg::PointFieldF32, offset);
  offset = addPointField(*message, "y", 1, beluga_ros::msg::PointFieldF32, offset);
  offset = addPointField(*message, "z", 1, beluga_ros::msg::PointFieldF32, offset);
  offset = addPointField(*message, "intensity", 1, beluga_ros::msg::PointFieldF32, offset);

  message->point_step = offset;
  message->row_step = message->width * message->point_step;
  message->is_dense = true;
  message->data.resize(message->point_step * message->width * message->height);

  auto cloud = beluga_ros::PointCloud3(message, origin);
  std::cout << cloud.points() << '\n';
  /*int i = 0;
  for (const auto& point : cloud.points()) {
    std::apply([&](auto x, auto y, auto z){
      ASSERT_EQ(x, point_data[3 * i + 0]);
      ASSERT_EQ(y, point_data[3 * i + 1]);
      ASSERT_EQ(z, point_data[3 * i + 2]);
    }, point);
    i++;
  }*/
}
```

Reply:
But without setting the fields it will **throw an exception**

```cpp
TEST(TestPointCloud, XYZIPointsEmptyFromUnorderedPointCloudMessage) {
  auto message = make_message();
  const auto origin = Sophus::SE3d{};
 

  auto cloud = beluga_ros::PointCloud3(message, origin);
  std::cout << cloud.points() << '\n';
  /*int i = 0;
  for (const auto& point : cloud.points()) {
    std::apply([&](auto x, auto y, auto z){
      ASSERT_EQ(x, point_data[3 * i + 0]);
      ASSERT_EQ(y, point_data[3 * i + 1]);
      ASSERT_EQ(z, point_data[3 * i + 2]);
    }, point);
    i++;
  }*/
}
```